R/type-list-of.R

```diff
@@ -188,11 +188,26 @@ vec_ptype2.vctrs_list_of <- function(x, y, ..., x_arg = "", y_arg = "") {
 }
 #' @method vec_ptype2.vctrs_list_of vctrs_list_of
 #' @export
-vec_ptype2.vctrs_list_of.vctrs_list_of <- function(x, y, ...) {
-  ptype <- vec_ptype2(attr(x, "ptype"), attr(y, "ptype"))
+vec_ptype2.vctrs_list_of.vctrs_list_of <- function(x, y, ..., x_arg = "", y_arg = "") {
+  frame <- current_env()
+  ptype <- tryCatch(
+    vec_ptype2(attr(x, "ptype"), attr(y, "ptype"), x_arg = x_arg, y_arg = y_arg),
+    vctrs_error_incompatible_type = function(cnd) {
+      return_from(frame, list())
+    }
```

Comment:
I'm a little worried about the performance of all this. Can we apply the same `identical()` check that we did in the cast method (see suggestion)? It seems to shave off about 20% or so. I can get the example down to 5 seconds by also providing a `vec_ptype.vctrs_list_of()` method (because right now it is doing `vec_slice(x, NULL)` on each input every time) but we can do that in another PR


```r
x <- list_of(.ptype = double())
xs <- rep(list(x), 1e5)

bench::mark(vec_ptype_common(!!!xs))

# Current PR
# A tibble: 1 × 13
  expression                   min median `itr/sec` mem_alloc `gc/sec` n_itr
  <bch:expr>              <bch:tm> <bch:>     <dbl> <bch:byt>    <dbl> <int>
1 vec_ptype_common(!!!xs)    10.3s  10.3s    0.0974     782KB     12.6     1
# … with 6 more variables: n_gc <dbl>, total_time <bch:tm>, result <list>,
#   memory <list>, time <list>, gc <list>

# With identical() check
# A tibble: 1 × 13
  expression                   min median `itr/sec` mem_alloc `gc/sec` n_itr
  <bch:expr>              <bch:tm> <bch:>     <dbl> <bch:byt>    <dbl> <int>
1 vec_ptype_common(!!!xs)     8.5s   8.5s     0.118     782KB     12.8     1
# … with 6 more variables: n_gc <dbl>, total_time <bch:tm>, result <list>,
#   memory <list>, time <list>, gc <list>
```

```suggestion
  x_ptype <- attr(x, "ptype", exact = TRUE)
  y_ptype <- attr(y, "ptype", exact = TRUE)

  if (identical(x_ptype, y_ptype)) {
    x
  } else {
    tryCatch(
      expr = {
        ptype <- vec_ptype2(x_ptype, y_ptype, x_arg = x_arg, y_arg = y_arg)
        new_list_of0(x = list(), ptype = ptype)
      },
      vctrs_error_incompatible_type = function(cnd) {
        list()
      }
    )
  }
```